haystack_experimental/core/pipeline/pipeline.py

```diff
@@ -229,45 +311,387 @@ def run(  # noqa: PLR0915, PLR0912
+                return transformed[0]
+            return transformed
+
+        else:
+            # For other data types, just return the value as is.
+            return data
+
+    @staticmethod
+    def _serialize_component_input(value: Any) -> Any:
+        """
+        Serializes, so it can be saved to a file, any type of input to a pipeline component.
+        """
+        value = Pipeline._remove_unserializable_data(value)
+        value = Pipeline.transform_json_structure(value)
+
+        if hasattr(value, "to_dict") and callable(getattr(value, "to_dict")):
+            serialized_value = value.to_dict()
+            serialized_value["_type"] = value.__class__.__name__
+            return serialized_value
+
+        # this is a hack to serialize inputs that don't have a to_dict
+        elif hasattr(value, "__dict__"):
+            return {
+                "_type": value.__class__.__name__,
+                "attributes": value.__dict__,
+            }
+
+        # recursively serialize all inputs in a dict
+        elif isinstance(value, dict):
+            return {k: Pipeline._serialize_component_input(v) for k, v in value.items()}
+
+        # recursively serialize all inputs in lists or tuples
+        elif isinstance(value, list):
+            return [Pipeline._serialize_component_input(item) for item in value]
+
+        return value
+
+    @staticmethod
+    def _deserialize_component_input(value: Any) -> Any:  # noqa: PLR0911
+        """
+        Tries to deserialize any type of input that can be passed to as input to a pipeline component.
+
+        For primitive values, it returns the value as is, but for complex types, it tries to deserialize them.
+        """
+
+        # None or primitive types are returned as is
+        if not value or isinstance(value, (str, int, float, bool)):
+            return value
+
+        # list of primitive types are returned as is
+        if isinstance(value, list) and all(isinstance(i, (str, int, float, bool)) for i in value):
+            return value
+
+        if isinstance(value, list):
+            # list of lists are called recursively
+            if all(isinstance(i, list) for i in value):
+                return [Pipeline._deserialize_component_input(i) for i in value]
+            # list of dicts are called recursively
+            if all(isinstance(i, dict) for i in value):
+                return [Pipeline._deserialize_component_input(i) for i in value]
+
+        # Define the mapping of types to their deserialization functions
+        _type_deserializers = {
+            "Answer": Answer.from_dict,
+            "ChatMessage": ChatMessage.from_dict,
+            "Document": Document.from_dict,
+            "ExtractedAnswer": ExtractedAnswer.from_dict,
+            "GeneratedAnswer": GeneratedAnswer.from_dict,
+            "SparseEmbedding": SparseEmbedding.from_dict,
+        }
```

Comment:
Similar to what we do for loading components couldn't we first import the `_type` check if it has a `from_dict` attribute and if so then call that? This way we could avoid needing a hard coded list right? [Here](https://github.com/deepset-ai/haystack/blob/0d6a3925068b0df821c295c1a8625911b72c38db/haystack/core/pipeline/base.py#L191-L215) is how we do this for loading components in `base.py`

Would also solve the case for custom components.